### PR TITLE
Add more features and clarification to Animators

### PR DESCRIPTION
**Pull request by Cadair**
This is a direct follow on from #3407 it adds the following functionality / cleanups:

* Use [`modest_image`](https://github.com/glue-viz/glue/blob/master/glue/external/modest_image.py) from Glue to improve rendering of images in `ArrayAnimatorWCS` by reading less of the array. (Especially useful with dask arrays or memmapped arrays).

* Allow passing through extra slider functions in the `ArrayAnimator` constructor, this lets you provide extra sliders for controlling the colourbar limits for instance.

* Make the specification of button labels optional, by default it will use the function names.

* Wire in *an existing keyword argument* to `BaseFuncAnimator` which allows the specification of a function to setup the plot as keyword argument. 

* Modify the call signature of the button call back functions so that they get passed the animator object, so that they can easily do things that modify the plot.

* Add optional slider labels to `BaseFuncAnimator` and set them by default in `ArrayAnimatorWCS` to the world axis names if present or the world axis physical types if not.

### CODE REVIEW

**Comment by DanRyanIrish** [comment]
Sounds potentially powerful.  Would have to see a demo to fully understand though.

**Comment by nabobalis** [comment]
Rebase conflicted for me (I suspect something is off with my repo) so I just merged in master.

File: sunpy/visualization/animator/tests/test_wcs.py [hunk 1]
```diff
@@ -126,6 +126,29 @@ def test_array_animator_wcs_2d_colorbar_buttons(wcs_4d):
...
+    def vmin_slider(val, im, slider):
+        im.set_clim(vmin=val)
+    def vmax_slider(val, im, slider):
+        im.set_clim(vmax=val)
+
+    data = np.arange(120).reshape((5, 4, 3, 2))
+    a = ArrayAnimatorWCS(data, wcs_4d, [0, 0, 'y', 'x'], colorbar=True,
+                         slider_functions=[vmin_slider, vmax_slider])
+    a.update_plot(1, a.im, a.sliders[0]._slider)
+    return a.fig
```
Comment (nabobalis) [inline sunpy/visualization/animator/tests/test_wcs.py]: Are these not missing the hashes?

**Comment by nabobalis** [comment]
Not sure how the figs failed when they passed on my machine?!

**Comment by Cadair** [comment]
There appears to be a codecov issue with uploading coverage on the py38-online build, so that test is not representative.

File: licenses/GLUE.rst [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Glue - multidimensional data exploration
```
Comment (dstansby) [inline licenses/GLUE.rst]: Maybe worth a note here referencing the file it is needed for?

File: sunpy/extern/modest_image.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
+"""
+Modification of Chris Beaumont's mpl-modest-image package to allow the use of
```
Comment (dstansby) [inline sunpy/extern/modest_image.py]: Has this gone upstream/if not is it worth going upstream?

File: sunpy/extern/modest_image.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    ModestImage is an extension of the Matplotlib AxesImage class
+    better suited for the interactive display of larger images. Before
+    drawing, ModestImage resamples the data array based on the screen
+    resolution and view window. This has very little affect on the
+    appearance of the image, but can substantially cut down on
+    computation since calculations of unresolved or clipped pixels
+    are skipped.
+
+    The interface of ModestImage is the same as AxesImage. However, it
+    does not currently support setting the 'extent' property. There
```
Comment (dstansby) [inline sunpy/extern/modest_image.py]: Change this bit of docstring, since this implementation does allow `extent`?

File: sunpy/visualization/animator/base.py [hunk 5]
```diff
@@ -52,22 +52,32 @@ class BaseFuncAnimator:
     colorbar: `bool`, optional
         Plot a colorbar. Defaults to `False`.
     button_labels: `list`, optional
-        A list of strings to label buttons. Defaults to `None`.
+        A list of strings to label buttons. Defaults to `None`, if not
```
Comment (dstansby) [inline sunpy/visualization/animator/base.py]:
```suggestion
        A list of strings to label buttons. Defaults to `None`. If `None`
```
I think this makes more sense

File: sunpy/visualization/animator/base.py [hunk 6]
```diff
@@ -52,22 +52,32 @@ class BaseFuncAnimator:
...
+        specified and ``button_func`` is specified, it will default to the
```
Comment (dstansby) [inline sunpy/visualization/animator/base.py]:
```suggestion
        and ``button_func`` is specified, it will default to the
```

File: sunpy/visualization/animator/base.py [hunk 7]
```diff
@@ -52,22 +52,32 @@ class BaseFuncAnimator:
...
+        names of the functions.
     button_func: `list`, optional
-        A list of functions to map to the buttons. Defaults to `None`.
+        A list of functions to map to the buttons, these functions are called
```
Comment (dstansby) [inline sunpy/visualization/animator/base.py]:
```suggestion
        A list of functions to map to the buttons. These functions are called
```

File: sunpy/visualization/animator/base.py [hunk 8]
```diff
@@ -52,22 +52,32 @@ class BaseFuncAnimator:
...
+        with two arguments, ``(animator, event)`` where the first argument is
+        the animator object, and the second is the matplotlib mouse event.
```
Comment (dstansby) [inline sunpy/visualization/animator/base.py]: Put an intersphinx link to `MouseEvent` in the mpl docs?

File: tox.ini [hunk 9]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py{36,37,38}-{offline,online,astropydev,numpydev,astropy31},build_docs,figure,conda
+envlist = py{36,37,38}-{offline,online,astropydev,numpydev,astropy31},build_docs,figure,figure_astropydev,conda
```
Comment (dstansby) [inline tox.ini]: Why this change?

File: sunpy/extern/modest_image.py (hunk 3 above)
Comment (Cadair) [reply sunpy/extern/modest_image.py]: That you would have to ask @astrofrog

File: sunpy/extern/modest_image.py (hunk 4 above)
Comment (Cadair) [reply sunpy/extern/modest_image.py]: I don't really want to change this file as it's in extern, I just want to be able to copy it out of Glue if there are updates there?

File: tox.ini (hunk 9 above)
Comment (Cadair) [reply tox.ini]: We forgot to add the new env to `tox.ini` in #3407

File: sunpy/visualization/animator/base.py [hunk 10]
```diff
@@ -434,10 +456,13 @@ def __init__(self, data, image_axes=[-2, -1], axis_ranges=None, **kwargs):
         for i in self.slider_axes:
             self.frame_slice[i] = 0
 
+        slider_functions = kwargs.pop("slider_functions", [])
+        slider_ranges = kwargs.pop("slider_ranges", [])
         base_kwargs = {
-            'slider_functions': [self.update_plot] * self.num_sliders,
-            'slider_ranges': [[0, dim] for dim in np.array(data.shape)[self.slider_axes]]
+            'slider_functions': ([self.update_plot] * self.num_sliders) + slider_functions,
```
Comment (DanRyanIrish) [inline sunpy/visualization/animator/base.py]: So does this mean the slider functions specified by the user in the ```slider_ranges``` kwargs are only extra slider functions, e.g. for a colourbar?  Data axes' sliders are still derived internally?

Same for ```slider_ranges```

File: sunpy/visualization/animator/wcs.py [hunk 11]
```diff
@@ -90,7 +91,27 @@ def __init__(self, data, wcs, slices, coord_params=None, ylim='dynamic', ylabel=
         self.ylim = ylim
         self.ylabel = ylabel
 
-        super().__init__(data, image_axes=image_axes, axis_ranges=None, **kwargs)
+        extra_slider_labels = []
+        if "slider_functions" in kwargs and "slider_labels" not in kwargs:
+            extra_slider_labels = [a.__name__ for a in kwargs['slider_functions']]
+
```
Comment (DanRyanIrish) [inline sunpy/visualization/animator/wcs.py]: Too many empty lines.

File: sunpy/extern/modest_image.py (hunk 4 above)
Comment (Cadair) [reply sunpy/extern/modest_image.py]: yeah, it means we don't have to depend on glue, which is a *large* dependency with a lot of qt and other things, really not worth it.

File: sunpy/visualization/animator/base.py (hunk 10 above)
Comment (Cadair) [reply sunpy/visualization/animator/base.py]: Yes, indeed it does.
